multisrc/overrides/mangathemesia/constellarscans/src/ConstellarScans.kt

```diff
@@ -1,7 +1,161 @@
+
+    override val client = super.client.newBuilder().addInterceptor { chain ->
+        val response = chain.proceed(chain.request())
+
+        val url = response.request.url
+        if (url.queryParameter(DESCRAMBLE) == null) {
+            return@addInterceptor response
+        }
+
+        val segments = url.pathSegments
+        val filenameWithoutExtension = segments.last().split(".")[0]
+        val fragment = segments[segments.lastIndex - 1]
+        val key = md5sum(fragment + filenameWithoutExtension)
+
+        val image = descrambleImage(response.body!!.byteStream(), key)
+        val body = image.toResponseBody("image/jpeg".toMediaTypeOrNull())
+        response.newBuilder()
+            .body(body)
+            .build()
+    }.build()
+
     override val seriesStatusSelector = ".status"
+
+    override fun pageListParse(document: Document): List<Page> {
+        val pageList = super.pageListParse(document)
+        return when {
+            document.selectFirst("script:containsData(_code)") != null -> descramblePageUrls(
+                pageList
+            )
+            document.selectFirst("script:containsData(ts_reader[_)") != null -> decodeDeviceLimitedChapter(
+                document
+            )
+            else -> pageList
+        }
+    }
+
+    private fun descramblePageUrls(pages: List<Page>): List<Page> {
+        return pages.map {
+            val lastSegment = it.imageUrl!!.split('/').last()
+            val filename = lastSegment
+                .replace(NOT_DIGIT_RE, "")
+                .chunked(2)
+                .joinToString("") { LOOKUP_STRING[it.toInt()].toString() }
+            Page(it.index, imageUrl = it.imageUrl!!.replace(lastSegment, filename))
+        }
+    }
+
+    private fun decodeDeviceLimitedChapter(document: Document): List<Page> {
+        val script = document.selectFirst("script:containsData(ts_reader[_)").data()
+        val fullKey = DESCRAMBLING_KEY_RE.find(script)?.groupValues?.get(1)
+            ?: throw Exception("Could not find suitable decryption key. Try opening the chapter again.")
+
+        val shiftBy = fullKey.substring(32..33).toInt(16)
+        val key = fullKey.substring(0..31) + fullKey.substring(34)
+
+        val fragmentAndImageCount = key.map {
+            var idx = LOOKUP_STRING_ALNUM.indexOf(it) - shiftBy
+            if (idx < 0) {
+                idx += LOOKUP_STRING_ALNUM.length
+            }
+            LOOKUP_STRING_ALNUM[idx]
+        }.joinToString("")
+        val fragment = fragmentAndImageCount.substring(0..31)
+        val imageCount = fragmentAndImageCount.substring(32).toInt()
+
+        val pages = mutableListOf<Page>()
+        for (i in 1..imageCount) {
+            pages.add(
+                Page(
+                    i,
+                    imageUrl = "$encodedUploadsPath/$fragment/${i.toString().padStart(5, '0')}.webp?$DESCRAMBLE=yes"
```

Comment:
```suggestion
                    imageUrl = "$encodedUploadsPath/$fragment/${i.toString().padStart(5, '0')}.webp#$DESCRAMBLE"
```

Use fragment so this tag doesn't get sent to the server. Use `url.fragment` in interceptor to check this part.

Another good alternative:

```kotlin
class DescrambleTag

override fun imageRequest(page: Page) =
    GET(page.imageUrl!!, headers).newBuilder()
        .tag(DescrambleTag::class.java, DescrambleTag())
        .build()

// in interceptor
if (request.tag(DescrambleTag::class.java) == null) return response
```